docs/blockstates/extendedStates.md

```diff
@@ -0,0 +1,82 @@
+
+Unlisted Properties
+-----------------------
+
+Unlisted properties are an alternative kind of `IProperty<V>`, where every possible value combination is *not* generated on startup. Because of this, and because they do not show up on the game's F3 debug menu, we call them "unlisted".
+The interface type for unlisted properties is `IUnlistedProperty<V>`. The interface methods for both types have nearly identical semantics, except that instead of `IProperty.getAllowedValues()`, which returns a collection of all valid values, we have `IUnlistedProperty.isValid(V)` to detect whether a value is valid.
+
+Examples:
+* Forge itself has a [`PropertyFloat`](https://github.com/MinecraftForge/MinecraftForge/blob/e3777f455991b6803eb826ff6305d17749462a3b/src/main/java/net/minecraftforge/common/property/PropertyFloat.java), which can hold any float value passing a certain Predicate.
+* Botania has a generic [`PropertyObject`](https://github.com/Vazkii/Botania/blob/c58ae21173ff63da2f107e328f19a31e0d609778/src/main/java/vazkii/botania/api/state/PropertyObject.java) used to pass any sort of `Object` on to a custom model.
+
+Declaring Unlisted Properties
+------------------------------------
+
+Unlisted properties are declared in `Block.createBlockState`, the same place as regular ("listed") properties. Instead of returning a `BlockStateContainer`, one must return an `ExtendedBlockState`. Forge provides a builder `BlockStateContainer.Builder`, which will automatically handle returning an `ExtendedBlockState` for you.
+
+Example:
+```Java
+@Override
+public BlockStateContainer createBlockState() {
+	return new BlockStateContainer.Builder(this).add(LISTED_PROP).add(UNLISTED_PROP).build();
+}
+```
+
+Filling Extended States
+----------------------------
+
+Before an `IBlockState` is used for rendering purposes, it will always have `Block.getExtendedState` called on it first. In this method, you will give all your unlisted properties values. Assuming you registered at least one unlisted property in the previous section, the `IBlockState` parameter can be safely casted to `IExtendedBlockState`, which has a `withProperty` method for unlisted properties analogous to its listed property cousin. Here, you can query whatever you want from the World, the Tile Entity, etc. (with appropriate safety checks, of course) and insert it into the extended blockstate.
+
+!!! Warning
+	It is highly recommended that your unlisted property values be immutable. Baked model implementations will use the extended state and unlisted values on multiple threads, so any value must be used in a threadsafe manner. The easiest way is to simply make that information an immutable snapshot.
+	Because of this, `Block.getExtendedState` is the last opportunity you have on the main thread to pass information to the rendering system. Anything you might possibly want to know in your custom `IBakedModel`, you should be passing an immutable snapshot of through `Block.getExtendedState`.
```

Comment:
This is not technically correct, `getExtendedState` is not strictly called on the main thread. The general advice is good here, but this particular part should be changed or omitted.

Reply:
I didn't know that. So does this mean we actually have zero methods of 100% thread-safely getting info to put into an extended state? That's a bit scary.

Reply:
Ah, not sure how I had this mixed up all along, but on the worker threads a chunkcache is used, which makes block access safe, but mutable TE state is still unsafe to access from getActualState and getExtendedState.

Reply:
It's threadsafe because getExtendedState is passed a ChunkCache and not a world. TEs are a tricky subject but you can write your own TE to be threadsafe, so perhaps worth a note.

Reply:
That's what I thought too, but a ChunkCache just aliases the Chunks from the parent world directly, meaning all unsychronized accesses to it from the cache and the world at the same time will race